docs: tw-1012-finality (#3537)
OffchainLabs/arbitrum-docs: docs/how-arbitrum-works/deep-dives/finality.mdx

---
title: Finality
description: 'Learn the fundamentals of finality on Arbitrum.'
author: pete-vielhaber
sme: Mehdi
user_story: As a current or prospective Arbitrum user, I want to learn more about finality.
content_type: concept
---

> Finality in <a data-quicklook-from="blockchain">blockchain</a> systems refers to the point at which a <a data-quicklook-from="transaction">transaction</a> becomes irreversible and is permanently included in the blockchain’s ledger. In the context of Arbitrum’s <a data-quicklook-from="arbitrum-nitro">Nitro</a> architecture, understanding finality is crucial for developers and users to make informed decisions about transaction confirmations, security guarantees, and application design.

Reviewer: ```suggestion
Finality in <a data-quicklook-from="blockchain">blockchain</a> systems refers to the point at which a <a data-quicklook-from="transaction">transaction</a> becomes irreversible and is permanently included in the blockchain’s ledger. In Arbitrum’s <a data-quicklook-from="arbitrum-nitro">Nitro</a> architecture, understanding finality is crucial for developers and users to make informed decisions about transaction confirmations, security guarantees, and application design.
```

## Levels of finality

1. **<a data-quicklook-from="soft-confirmation">Soft finality</a>**: Provided by the <a data-quicklook-from="sequencer">Sequencer’s</a> real-time feed, offering immediate but provisional transaction confirmations.

> 2. **<a data-quicklook-from="hard-finality">Hard finality</a>**: Occurs when transactions are included in <a data-quicklook-from="batch">batches</a> posted to and finalized on the <a data-quicklook-from="parent-chain">parent chain</a>, providing strong security assurances. Final settlement on the parent chain happens via <a data-quicklook-from="assertion">assertions</a> confirmed by the <a data-quicklook-from="arbitrum-rollup-protocol">Rollup protocol</a>.

Reviewer: ```suggestion
2. **<a data-quicklook-from="hard-finality">Hard finality</a>**: Occurs when transactions are included in <a data-quicklook-from="batch">batches</a> posted to and finalized on the <a data-quicklook-from="parent-chain">parent chain</a>, providing strong security assurances. Final settlement on the <a data-quicklook-from="parent-chain">parent chain</a> happens via <a data-quicklook-from="assertion">assertions</a> confirmed by the <a data-quicklook-from="arbitrum-rollup-protocol">Rollup protocol</a>.
```

> This section explores the concepts of soft and hard finality, their implications, trust considerations, and guidance for using them effectively within the Arbitrum network.

Reviewer: ```suggestion
This section explores the concepts of soft and <a data-quicklook-from="hard-finality">hard finality</a>, their implications, trust considerations, and guidance for using them effectively within the Arbitrum network.
```

## Authority and finality

> - **Exclusive access**: Only authorized [<a data-quicklook-from="batch-poster">batch posters</a>](/how-arbitrum-works/deep-dives/sequencer.mdx#batch-posting) can call these batch-submission methods on the <a data-quicklook-from="sequencer-inbox">Sequencer Inbox</a> contract (access is gated by the contract's `isBatchPoster` mapping, a role distinct from the sequencer, though the two are typically operated by the same party). This exclusivity ensures that no other party can include messages directly.

Reviewer: ```suggestion
- **Exclusive access**: Only authorized [<a data-quicklook-from="batch-poster">batch posters</a>](/how-arbitrum-works/deep-dives/sequencer.mdx#batch-posting) can call these batch-submission methods on the <a data-quicklook-from="sequencer-inbox">Sequencer Inbox</a> contract (access is gated by the contract’s `isBatchPoster` mapping, a role distinct from the sequencer, though the two are typically operated by the same party). This exclusivity ensures that no other party can include messages directly.
```

- **Soft-confirmation receipts**: The Sequencer’s unique ability to immediately process and include transactions allows it to provide users with instant, “soft-confirmation” receipts.
- **Parent chain finality**: Once batches post, the transactions achieve parent-chain-level finality, secured by the parent chain’s consensus mechanism.

> By efficiently sending compressed transaction batches to the Sequencer Inbox contract using the most cost-effective method available, the Sequencer ensures transactions are securely recorded on the parent chain. This process maintains the integrity and reliability of the network, enabling users to perform fast, secure transactions.

Reviewer: ```suggestion
By efficiently sending compressed transaction batches to the <a data-quicklook-from="sequencer-inbox">Sequencer Inbox</a> contract using the most cost-effective method available, the Sequencer ensures transactions are securely recorded on the parent chain. This process maintains the integrity and reliability of the network, enabling users to perform fast, secure transactions.
```

| Stage                                                                                         | Guarantee                                                                                   | Approximate latency                                                                                                                                    |
| --------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------ |
| Soft finality (<a data-quicklook-from="sequencer-feed">Sequencer feed</a>)                    | Provisional ordering by the Sequencer                                                       | Sub-second                                                                                                                                             |

> | Parent-chain data finality (batch posted and finalized on <a data-quicklook-from="L1">L1</a>) | Transaction data is durably recorded and inherits the parent chain's consensus security     | ~2 epochs, roughly 12–15 minutes (the batch poster defaults to referencing the L1 `safe` block)                                                        |

Reviewer: ```suggestion
| Parent-chain data finality (batch posted and finalized on <a data-quicklook-from="L1">L1</a>) | Transaction data is durably recorded and inherits the parent chain's consensus security     | ~2 epochs, roughly 12–15 minutes (the <a data-quicklook-from="batch-poster">batch poster</a> defaults to referencing the L1 `safe` block)                                                        |
```

> | Full settlement (assertion confirmed)                                                         | <a data-quicklook-from="L2">L2</a> state is finalized and L2→L1 withdrawals can be executed | A bounded <a data-quicklook-from="challenge-period">challenge period</a> under <a data-quicklook-from="bold">BoLD</a>, on the order of days on mainnet |

Reviewer: ```suggestion
| Full settlement (<a data-quicklook-from="assertion">assertion</a> confirmed)                                                         | <a data-quicklook-from="L2">L2</a> state is finalized and L2→L1 withdrawals can be executed | A bounded <a data-quicklook-from="challenge-period">challenge period</a> under <a data-quicklook-from="bold">BoLD</a>, on the order of days on mainnet |
```

- **Immediate confirmation**: Transactions are confirmed almost instantly as they are accepted and ordered by the Sequencer.
- **Provisional assurance**: The confirmations are provisional and rely on the Sequencer’s integrity and availability.
- **High performance**: Enables applications to offer rapid responses and real-time interactions, enhancing user experience.

### Advantages of soft finality

- **Low latency**: Users receive immediate feedback on transaction status.
- **Optimized for speed**: Ideal for applications where responsiveness is critical.
- **Improved user experience**: Reduces waiting times and uncertainty.

### Limitations of soft finality

- **Trust dependency**: Relies on the Sequencer’s honesty and ability to maintain uptime.
- **Potential for reordering**: In rare cases, if the Sequencer acts maliciously or encounters issues, the provisional ordering could change.
- **Not suitable for high-value transactions**: For transactions requiring strong security guarantees, soft finality may not suffice.

## Censorship resistance and force inclusion

> The soft-finality trust dependency on the Sequencer is bounded rather than open-ended. A user can submit a transaction directly to the parent chain through the <a data-quicklook-from="delayed-inbox">delayed inbox</a>. If the Sequencer refuses to include it, the user can call <a data-quicklook-from="forceinclusion">`forceInclusion`</a> on the Sequencer Inbox contract once the Sequencer's exclusive window has elapsed, pushing the message into the chain without the Sequencer's cooperation (see [censorship resistance and the Censorship Timeout](/how-arbitrum-works/deep-dives/sequencer.mdx#censorship-timeout)). By default this window is roughly 24 hours (configured as `delayBlocks`/`delaySeconds`, defaulting to about 5,760 blocks or 86,400 seconds).

Reviewer: ```suggestion
The soft-finality trust dependency on the Sequencer is bounded rather than open-ended. A user can submit a transaction directly to the parent chain through the <a data-quicklook-from="delayed-inbox">delayed inbox</a>. If the Sequencer refuses to include it, the user can call <a data-quicklook-from="forceinclusion">`forceInclusion`</a> on the Sequencer Inbox contract once the Sequencer’s exclusive window has elapsed, pushing the message into the chain without the Sequencer’s cooperation (see [censorship resistance and the Censorship Timeout](/how-arbitrum-works/deep-dives/sequencer.mdx#censorship-timeout)). By default this window is roughly 24 hours (configured as `delayBlocks`/`delaySeconds`, defaulting to about 5,760 blocks or 86,400 seconds).
```

> The practical consequence is that the Sequencer can _delay_ a transaction but cannot _permanently_ censor it. Delayed-inbox messages are only sequenced after they reach parent-chain finality — the delayed sequencer waits for the parent chain's `safe` (or, if configured, `finalized`) head before including them — so force-included transactions inherit parent-chain finality guarantees.

Reviewer: ```suggestion
The practical consequence is that the Sequencer can _delay_ a transaction but cannot _permanently_ censor it. Delayed-inbox messages are only sequenced after they reach parent-chain finality — the delayed sequencer waits for the parent chain’s `safe` (or, if configured, `finalized`) head before including them — so force-included transactions inherit parent-chain finality guarantees.
```

## Hard finality

> Hard finality occurs when batched transactions are posted to the parent chain and the corresponding assertions are <a data-quicklook-from="confirmation">confirmed</a> by the Rollup protocol. Note that posting a batch to the parent chain is not itself final settlement: the transaction data becomes available and inherits the parent chain's data-layer security, but final settlement of the resulting L2 state occurs once assertions are confirmed (see "Final settlement" above). Key characteristics include:

Reviewer: ```suggestion
Hard finality occurs when batched transactions are posted to the parent chain and the corresponding assertions are <a data-quicklook-from="confirmation">confirmed</a> by the Rollup protocol. Posting a batch to the parent chain is not itself final settlement: the transaction data becomes available and inherits the parent chain’s data-layer security, but final settlement of the resulting L2 state occurs once assertions are confirmed (see "Final settlement" above). Key characteristics include:
```

- **Irreversibility**: Once finalized, transactions are immutable and cannot be altered or reversed.
- **Data availability**: In the default [Rollup mode](/run-arbitrum-node/data-availability.mdx#rollup-mode), all transaction data is recorded onchain (via calldata or <a data-quicklook-from="blob">blobs</a>), ensuring transparency and verifiability. On <a data-quicklook-from="arbitrum-anytrust-chain">AnyTrust chains</a>, batch data is instead kept offchain by a <a data-quicklook-from="data-availability-committee-dac">Data Availability Committee (DAC)</a>, and only a <a data-quicklook-from="data-availability-certificate">data availability certificate</a> is posted onchain — trading full onchain data availability for lower cost.

### Advantages of hard finality

- **Maximum security**: Protected by the robustness of the parent chain’s consensus mechanism.
- **Trust minimization**: This does not require trust in the Sequencer; the underlying blockchain provides security.
- **Suitable for high-value transactions**: Ideal for scenarios where security and immutability are paramount.

### Limitations of hard finality

> - **High latency**: Achieving hard finality takes longer because the parent chain must process and finalize batches. The <a data-quicklook-from="challenge-protocol">challenge protocol</a> that backs this guarantee is described in BoLD.

Reviewer: ```suggestion
- **High latency**: Achieving hard finality takes longer because the parent chain must process and finalize batches. The <a data-quicklook-from="challenge-protocol">challenge protocol</a> that backs this guarantee is described in <a data-quicklook-from="bold">BoLD</a>.
```

> - **Cost considerations**: Posting batches to the parent chain incurs fees, potentially increasing transaction costs.

Reviewer 2: While this is true, labeling it as a limitation seems odd since no matter what your transaction will be posted via the batch poster and you will pay for its fees, its not something that only happens with hard finality. 
```suggestion

```

> - **Cost considerations**: Posting batches to the parent chain incurs fees, potentially increasing transaction costs.

Author: ```suggestion
```

### Hard finality trust model

- **Reliance on the parent chain**: Security is based on the consensus and integrity of the parent chain.
- **Reduced trust in the Sequencer**: Even if the Sequencer misbehaves, transactions included in posted batches are secured once finalized on the parent chain.

## Application implications

Developers and users should consider the appropriate level of finality based on their specific use cases:

### When to rely on soft finality

- **Low-risk transactions**: For transactions where the potential impact of reordering or delays is minimal.
- **User experience priority**: Applications where responsiveness and immediacy enhance user engagement, such as gaming or social platforms.
- **Frequent transactions**: Scenarios involving a high volume of small transactions where waiting for hard finality is impractical.

### When to require hard finality

- **High-value transactions**: Financial transfers, large trades, or any transaction where security is critical.
- **Regulatory compliance**: Situations requiring strict adherence to security standards and auditable records.

> - **Centralized exchanges** (CEXs): For deposit and withdrawal operations where certainty of transaction finality is mandatory. Note that these two operations rely on different guarantees. Crediting a deposit can rely on parent-chain data finality (batches posted and finalized on L1, on the order of minutes). Releasing an <a data-quicklook-from="l2-to-l1-message">L2→L1 withdrawal</a>, however, requires the corresponding assertion to be confirmed by the Rollup protocol, which only completes after the BoLD challenge period (on the order of days). Batch posting alone does not make funds withdrawable to the parent chain.

Reviewer: ```suggestion
- **Centralized exchanges** (CEXs): For deposit and withdrawal operations where certainty of transaction finality is mandatory. These two operations rely on different guarantees. Crediting a deposit can rely on parent-chain data finality (batches posted and finalized on L1, on the order of minutes). Releasing an <a data-quicklook-from="l2-to-l1-message">L2→L1 withdrawal</a>, however, requires the corresponding assertion to be confirmed by the Rollup protocol, which only completes after the BoLD challenge period (on the order of days). Batch posting alone does not make funds withdrawable to the parent chain.
```

## Related resources

> - [Finality and chain reorganizations](/how-arbitrum-works/reference/finality-and-reorgs.mdx) — the `latest`, `safe`, and `finalized` block tags, reorg depth, and which tag an indexer should wait on.

Reviewer: ```suggestion
- [Finality and chain reorganizations](/how-arbitrum-works/reference/finality-and-reorgs.mdx) — the `latest`, `safe`, and `finalized` block tags, <a data-quicklook-from="reorg">reorg</a> depth, and which tag an indexer should wait on.
```

> - [The Sequencer and censorship resistance](/how-arbitrum-works/deep-dives/sequencer.mdx) — the real-time feed, batch posting, and the delayed inbox escape hatch.

Reviewer: ```suggestion
- [The Sequencer and censorship resistance](/how-arbitrum-works/deep-dives/sequencer.mdx) — the real-time feed, batch posting, and the <a data-quicklook-from="delayed-inbox">delayed inbox</a> escape hatch.
```

> - [Assertions](/how-arbitrum-works/deep-dives/assertions.mdx) — how the Rollup protocol confirms child chain state and completes final settlement.

Reviewer: ```suggestion
- [Assertions](/how-arbitrum-works/deep-dives/assertions.mdx) — how the Rollup protocol confirms <a data-quicklook-from="child-chain">child chain</a> state and completes final settlement.
```

- [BoLD: a gentle introduction](/how-arbitrum-works/bold/gentle-introduction.mdx) — the dispute protocol that bounds the challenge period.

> - [AnyTrust protocol](/how-arbitrum-works/deep-dives/anytrust-protocol.mdx) — how a DAC and data availability certificates change the data availability guarantee.

Reviewer: ```suggestion
- [AnyTrust protocol](/how-arbitrum-works/deep-dives/anytrust-protocol.mdx) — how a DAC and <a data-quicklook-from="data-availability-certificate">data availability certificates</a> change the data availability guarantee.
```

> - [Child to parent chain messaging](/how-arbitrum-works/deep-dives/l2-to-l1-messaging.mdx) — the outbox and why withdrawals wait for assertion confirmation.

Reviewer: ```suggestion
- [Child to parent chain messaging](/how-arbitrum-works/deep-dives/l2-to-l1-messaging.mdx) — the <a data-quicklook-from="outbox">outbox</a> and why withdrawals wait for assertion confirmation.
```
